pkg/v1/layout/puller.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2024 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package layout
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+
+	"github.com/google/go-containerregistry/pkg/v1/partial"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	specsv1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type puller struct {
+	path Path
+}
+
+func NewPuller(path Path) remote.Puller {
```

Comment:
I think we want to define a new interface in `v1` that both `remote.Puller` and `layout.Source` would implement.

In my mind I'd want:

```
package v1

type Source interface{
 // minimal subset of methods needed to read things
}

type Sink interface{
 // minimal subset of methods needed to write things

}
```

Reply:
That would remove the dependency between `layout` + `remote` and avoid the need to convert `remote.Puller` into a struct below. Spitballing the set of methods:

```go
type Source interface {
  Layer(...)
  Head(...)
  Get(...)
  Artifact(...)
}

type Sink interface {
  Push(...)
  Upload(...) // Maybe not?
}
```

The verbs are a bit clunky for non-remote implementors, but I don't see a clean way around that.

Reply:
Hmm, i can't put source and sink into v1 root folder as it creates cycle between `v1` -> `partial.Artifact` -> `v1`


Would putting this into its own directory be acceptable?

```
package github.com/google/go-containerregistry/cmd/crane
        imports github.com/google/go-containerregistry/cmd/crane/cmd
        imports github.com/google/go-containerregistry/internal/cmd
        imports github.com/google/go-containerregistry/internal/verify
        imports github.com/google/go-containerregistry/pkg/v1
        imports github.com/google/go-containerregistry/pkg/v1/partial
        imports github.com/google/go-containerregistry/pkg/v1: import cycle not allowed
```

Reply:
okay i have created the sinksource (name TBD) directory.

Reply:
This is now ready for another look

Reply:
@jonjohnsonjr what do you think about `v1.Artifact` to get around the import cycle?

Reply:
that one also sounds fair, though now v1.Artifact would depend on bunch of partial interfaces.

Reply:
anything i could do to push this over the edge @jonjohnsonjr? I'd really like to land this, as i have some people who's been waiting for it for a long time.